cpp/include/cudf/dfa/superstate.hpp

```diff
@@ -0,0 +1,224 @@
+  static const uint8_t BITS = 3;
+  static const uint8_t MASK = 0b111;
+
+  using Data = uint16_t;
+};
+
+template <uint8_t N>
+struct superstate_policy<N, typename std::enable_if<6 <= N and N <= 8>::type> {
+  static const uint8_t BITS = 3;
+  static const uint8_t MASK = 0b111;
+
+  using Data = uint32_t;
+};
+
+template <uint8_t N>
+struct superstate_policy<N, typename std::enable_if<9 <= N and N <= 16>::type> {
+  static const uint8_t BITS = 4;
+  static const uint8_t MASK = 0b1111;
+
+  using Data = uint64_t;
+};
+
+}  // namespace
+
+/**
+ * @brief Represents all possible dfa states and the transitions applied to them
+ *
+ * Used to compute dfa states and transitions in parallel. Stores states in a compressed format.
+ *
+ * Requires `N * log2(N)` storage bits:
+ * - up to  4 states + all transitions in 1 byte  (uint8_t)
+ * - up to  5 states + all transtiions in 2 bytes (uint16_t)
+ * - up to  8 states + all transitions in 4 bytes (uint32_t)
+ * - up to 16 states + all transitions in 8 bytes (uint64_t)
+ *
+ * ```
+ * // alias superstate for convienence.
+ * // csv_state has 7 unique states, and will therefore require a total of 4 bytes (uint32_t)
+ * using csv_superstate = superstate<csv_state, csv_token, 7>;
+ *
+ * // apply state transitions in "parallel"
+ * auto a = csv_superstate() + csv_token::comment + csv_token::other + csv_token::newline;
+ * auto b = csv_superstate() + csv_token::other + csv_token::delimiter;
+ * auto c = csv_superstate() + csv_token::other + csv_token::delimiter;
+ *
+ * // given record_end as the initial state, get the resulting state after all transitions.
+ * auto final = a + b + c;
+ * auto result_a = final.get(csv_state::record_end);   // csv_state::field_end
+ * auto result_b = final.get(csv_state::quoted_field); // csv_state::quoted_field
+ * auto result_c = final.get(csv_state::field_end);    // csv_state::invalid
+ * ```
+ *
+ * @note this class is constexpr, and can be used on both host and device
+ *
+ * @tparam State state which will be tracked - up to 16 unique values. usually an enum class.
+ * @tparam Instruction type which can be added to a state to transition in to a new state.
+ * @tparam N total number of states representable by State, up to 16 unique states.
+ */
+template <typename State, typename Instruction, uint8_t N>
+struct superstate {
+ private:
+  static const uint8_t BITS = superstate_policy<N>::BITS;
+  static const uint8_t MASK = superstate_policy<N>::MASK;
+
+  using Data = typename superstate_policy<N>::Data;
+
+  Data _data;
+
+ public:
+  /**
+   * @brief creates a superstate which represents all possible states and applied transitions
+   */
+  inline constexpr superstate() : _data(0)
+  {
+    for (auto i = 0; i < N; i++) { _data |= i << (i * BITS); }
+  }
+
+  /**
+   * @brief creates a superstate which represents only a single state
+   */
+  inline constexpr superstate(State state) : _data(0)
```

Comment:
Careful with implicit conversion ctors. They can cause a lot of headache. Consider making explicit.